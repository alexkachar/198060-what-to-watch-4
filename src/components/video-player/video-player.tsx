import * as React from 'react';
import {Link} from 'react-router-dom';
import {connect} from 'react-redux';
import {compose} from 'recompose';

import {formatTime} from '../../utils';
import Movie from '../../interfaces/movie';
import {getMovieById} from '../../store/reducers/ui/selectors';
import UiActionCreator from '../../store/actions/ui/ui';

import FullScreenButton from '../partials/full-screen-button/full-screen-button';
import PlayButton from '../partials/play-button/play-button';
import Loader from '../loader/loader';
import withPlayerControls from '../../hocs/with-player-controls/with-player-controls';

interface Props {
  movieId: string;
  movie: Movie;
  progress: number;
  timeLeft: number;
  isPlaying: boolean;
  onPlayButtonClick: () => void;
  onFullScreenButtonClick: () => void;
  onExitButtonClick: () => void;
  onSetDuration: () => void;
  onTimeUpdate: () => void;
  onSetMovieId: (movieId: number | string) => void;
}

class VideoPlayer extends React.PureComponent<Props> {
  private _videoRef;
  private _playerRef;

  constructor(props) {
    super(props);

    this._videoRef = React.createRef();
    this._playerRef = React.createRef();

    this._handleFullScreenButtonClick = this._handleFullScreenButtonClick.bind(this);
  }

  componentDidMount() {
    const {onSetMovieId, movieId} = this.props;
    onSetMovieId(movieId);
  }

  componentDidUpdate(prevProps) {
    const video = this._videoRef.current;
    const {onSetMovieId, movieId, isPlaying} = this.props;

    if (movieId !== prevProps.offerId) {
      onSetMovieId(movieId);
    }

    if (isPlaying) {
      video.play();
    } else {
      video.pause();
    }
  }

  _handleFullScreenButtonClick() {
    const player = this._playerRef.current;

    if (!document.fullscreenElement) {
      player.requestFullscreen();
    } else {
      document.exitFullscreen();
    }
  }

  render() {
    if (!this.props.movie) {
      return <Loader />;
    }

    const {
      isPlaying,
      progress,
      timeLeft,
      onPlayButtonClick,
      onExitButtonClick,
      onSetDuration,
      onTimeUpdate,
      movie
    } = this.props;

    const {
      title,
      backgroundImage,
      videoLink
    } = movie;

    return (
      <div className="player" ref={this._playerRef}>
        <video
          className="player__video"
          ref={this._videoRef}
          poster={backgroundImage}
          src={videoLink}
          onCanPlay={onSetDuration}
          onTimeUpdate={onTimeUpdate}
        />
        <Link to="/">
          <button
            type="button"
            className="player__exit"
            onClick={onExitButtonClick}
          >
            Exit
          </button>
        </Link>
        <div className="player__controls">
          <div className="player__controls-row">
            <div className="player__time">
              <progress className="player__progress" value={progress} max={100} />
              <div className="player__toggler" style={{left: `${progress}%`}}>Toggler</div>
            </div>
            <div className="player__time-value">{formatTime(timeLeft)}</div>
          </div>
          <div className="player__controls-row">
            <PlayButton
              isPlaying={isPlaying}
              onPlayButtonClick={onPlayButtonClick}
            />
            <div className="player__name">{title}</div>
            <FullScreenButton onFullScreenButtonClick={this._handleFullScreenButtonClick} />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  movie: getMovieById(state),
});

const mapDispatchToProps = (dispatch) => ({

  onSetMovieId: (movieId) => {
    const newMovieId = parseInt(movieId, 10);
    dispatch(UiActionCreator.setMovieId(newMovieId));
  }

});

const withPlayersControlsAndConnect = compose(
    connect(mapStateToProps, mapDispatchToProps),
    withPlayerControls
);

export {VideoPlayer};

export default withPlayersControlsAndConnect(VideoPlayer);
