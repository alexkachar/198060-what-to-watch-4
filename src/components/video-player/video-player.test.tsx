import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {BrowserRouter} from 'react-router-dom';
import {VideoPlayer} from './video-player';

import mockMovies from '../../test-data/mock-movies';

it(`VideoPlayer renders correctly`, () => {
  const tree = renderer
    .create(
        <BrowserRouter>
          <VideoPlayer
            movieId={`1`}
            movie={mockMovies[0]}
            progress={30}
            timeLeft={70}
            isPlaying={true}
            onPlayButtonClick={jest.fn()}
            onFullScreenButtonClick={jest.fn()}
            onExitButtonClick={jest.fn()}
            onSetDuration={jest.fn()}
            onTimeUpdate={jest.fn()}
            onSetMovieId={jest.fn()}
          />
        </BrowserRouter>, {
          createNodeMock: () => document.createElement(`div`)
        })
    .toJSON();
  expect(tree).toMatchSnapshot();
});
