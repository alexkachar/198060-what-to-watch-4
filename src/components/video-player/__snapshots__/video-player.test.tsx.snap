// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPlayer renders correctly 1`] = `
<div
  className="player"
>
  <video
    className="player__video"
    onCanPlay={[MockFunction]}
    onTimeUpdate={[MockFunction]}
    poster="img/the-grand-budapest-hotel-bg.jpg"
    src="https://some-link"
  />
  <a
    href="/"
    onClick={[Function]}
  >
    <button
      className="player__exit"
      onClick={[MockFunction]}
      type="button"
    >
      Exit
    </button>
  </a>
  <div
    className="player__controls"
  >
    <div
      className="player__controls-row"
    >
      <div
        className="player__time"
      >
        <progress
          className="player__progress"
          max={100}
          value={30}
        />
        <div
          className="player__toggler"
          style={
            Object {
              "left": "30%",
            }
          }
        >
          Toggler
        </div>
      </div>
      <div
        className="player__time-value"
      >
        00:01:10
      </div>
    </div>
    <div
      className="player__controls-row"
    >
      <button
        className="player__play"
        onClick={[MockFunction]}
        type="button"
      >
        <svg
          height={19}
          viewBox="0 0 19 19"
          width={19}
        >
          <use
            xlinkHref="#pause"
          />
        </svg>
        <span>
          Play
        </span>
      </button>
      <div
        className="player__name"
      >
        The Grand Budapest Hotel
      </div>
      <button
        className="player__full-screen"
        onClick={[Function]}
        type="button"
      >
        <svg
          height={27}
          viewBox="0 0 27 27"
          width={27}
        >
          <use
            xlinkHref="#full-screen"
          />
        </svg>
        <span>
          Full screen
        </span>
      </button>
    </div>
  </div>
</div>
`;
